packages/background/src/controllers/NetworkController.ts

```
@@ -781,6 +831,101 @@ export default class NetworkController extends BaseController<NetworkControllerS
+     *
+     */
+    public async switchProvider(
+        chainId: number,
+        providerType: ProviderType,
+        customRpcUrl?: string
+    ): Promise<void> {
+        const network = this.getNetworkFromChainId(chainId);
+        if (!network) throw new Error('Network not found for given chainId');
+
+        let rpcUrl;
+        switch (providerType) {
+            case ProviderType.DEFAULT:
+                rpcUrl = network.defaultRpcUrl;
+                break;
+            case ProviderType.BACKUP: {
+                rpcUrl = await this._getWorkingBackupRpcUrl(chainId);
+                break;
+            }
+            case ProviderType.CUSTOM:
+                rpcUrl = customRpcUrl;
+                break;
+            default:
+                throw new Error('Invalid provider type');
+        }
+        if (!rpcUrl) throw new Error('Invalid rpc url');
+
+        await this.editNetwork(chainId, {
+            rpcUrls: [rpcUrl],
+            blockExplorerUrls: network.blockExplorerUrls,
+            name: network.desc,
+            test: network.test,
+        });
+
+        this.store.updateState({
+            providerStatus: {
+                ...this.getState().providerStatus,
+                isUsingBackupProvider: providerType === ProviderType.BACKUP,
+            },
+        });
+    }
+
+    /**
+     *  _isRpcValid
+     *
+     *  Checks if the rpc url is working
+     *  @param chainId chain identifier of the network
+     *  @param rpcUrl rpc url of the network
+     *  @returns true if the rpc url is valid, false otherwise
+     */
+    private async _isRpcValid(
+        chainId: number,
+        rpcUrl: string
+    ): Promise<boolean> {
+        let isValid = false;
+        try {
+            const returnedChainId = await getCustomRpcChainId(rpcUrl);
+            if (returnedChainId === chainId) {
+                isValid = true;
+            }
+        } catch (error) {
+            // isValid is false
```

can you add a `log.warn` here? It would be good to have the error the rpc is throwing

---

even if the error is because the RPC is not working which is the intended behaviour?

---

You're right. No need to log here. I was worried about the empty catch... but as you mentioned above you will do: `return false;`